File: 개인문제/편근형/유기농 배추.java
// 링크 : https://www.acmicpc.net/problem/1012
// 시간 :184ms
// 코드 길이 : 1959B
// 메모리 : 15784KB

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.StringTokenizer;
import java.util.Stack;

public class Main {
  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer(br.readLine(), " ");
    int t = Integer.parseInt(st.nextToken());

    for (int a = 0; a < t; a++) {
      st = new StringTokenizer(br.readLine(), " ");
      int m = Integer.parseInt(st.nextToken());
      int n = Integer.parseInt(st.nextToken());
      int k = Integer.parseInt(st.nextToken());
      int[][] direction = {{1,0}, {-1,0}, {0,1}, {0,-1}};
      Stack<int[]> stack = new Stack<>();
      int[][] field = new int[n][m]; // 선언과 배열을 0으로 초기화 동시에 진행
      int cnt = 0;
  
      for (int i = 0; i < k; i++) {
        st = new StringTokenizer(br.readLine(), " ");
        int x = Integer.parseInt(st.nextToken());
        int y = Integer.parseInt(st.nextToken());
        field[y][x] = 1;
      }
  
      for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
          if(field[i][j] == 1) {
            cnt += 1;
            field[i][j] = 0;
            int[] point = {i,j};
            stack.add(point);
          }
          while (stack.empty() == false) {
            int[] temp = stack.pop();
            int x = temp[1]; int y = temp[0];
  
            for(int[] dir : direction) {
              int dx = dir[1]; int dy = dir[0];
              int nx = dx + x;
              int ny = dy + y;
              if (nx >= 0 && nx < m && ny >= 0 && ny < n) {
                if (field[ny][nx] == 1) {
                  field[ny][nx] = 0;
                  int[] newPoint = {ny, nx};
                  stack.push(newPoint);
                }
              }
            }
          }
        }
      }
      System.out.println(cnt);
    }   
  }
}

File: 개인문제/편근형/유기농 배추.py
# 링크 : https://www.acmicpc.net/problem/1012
# 시간 : 88ms
# 코드 길이 : 734B

import sys
ssr = sys.stdin.readline

t = int(ssr())
direction = [(0,1),(1,0),(0,-1),(-1,0)]
for _ in range(t):
  m,n,k = list(map(int, ssr().split(' ')))
  cnt = 0
  field = [[0 for _ in range(m)] for _ in range(n)]
  stack = []
  for _ in range(k):
    x, y = list(map(int, ssr().split(' ')))
    field[y][x] = 1

  for i in range(n):
    for j in range(m):
      if field[i][j] == 1:
        cnt += 1
        stack.append((i,j))
        field[i][j] = 0
      while stack:
        y,x = stack.pop()
        for dy, dx in direction:
          nx = dx + x
          ny = dy + y
          if 0 <= nx < m and 0 <= ny < n:
            if field[ny][nx] == 1:
              field[ny][nx] = 0
              stack.append((ny,nx))
  print(cnt)
